keydata.go

```diff
@@ -932,9 +951,121 @@ func NewKeyData(params *KeyParams) (*KeyData, error) {
 	return kd, nil
 }
 
+func NewKeyDataWithPassphrase(params *KeyWithPassphraseParams, passphrase string, kdf KDF) (*KeyData, error) {
+	kd, err := NewKeyData(&params.KeyParams)
+	if err != nil {
+		return nil, err
+	}
+
+	kdfOptions := params.KDFOptions
+	if kdfOptions == nil {
+		var defaultOptions KDFOptions
+		kdfOptions = &defaultOptions
+	}
+
+	costParams, err := kdfOptions.deriveCostParams(passphraseEncryptionKeyLen+aes.BlockSize, kdf)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot derive KDF cost parameters: %w", err)
+	}
+
+	var salt [16]byte
+	if _, err := rand.Read(salt[:]); err != nil {
+		return nil, xerrors.Errorf("cannot read salt: %w", err)
+	}
+
+	kd.data.PassphraseParams = &passphraseParams{
+		KDF: kdfData{
+			Type:   kdfType,
+			Salt:   salt[:],
+			Time:   int(costParams.Time),
+			Memory: int(costParams.MemoryKiB),
+			CPUs:   int(costParams.Threads),
+		},
+		Encryption:        passphraseEncryption,
+		DerivedKeySize:    passphraseKeyLen,
+		EncryptionKeySize: passphraseEncryptionKeyLen,
+		AuthKeySize:       params.AuthKeySize,
+	}
+
+	if err := kd.updatePassphrase(kd.data.EncryptedPayload, make([]byte, params.AuthKeySize), passphrase, kdf); err != nil {
+		return nil, xerrors.Errorf("cannot set passphrase: %w", err)
+	}
+
+	return kd, nil
+}
+
+type protectedKeys struct {
+	Primary PrimaryKey
+	Unique  []byte
+}
+
+func unmarshalProtectedKeys(data []byte) (*protectedKeys, error) {
+	s := cryptobyte.String(data)
+	if !s.ReadASN1(&s, cryptobyte_asn1.SEQUENCE) {
+		return nil, errors.New("malformed input")
+	}
+
+	pk := new(protectedKeys)
+
+	if !s.ReadASN1Bytes((*[]byte)(&pk.Primary), cryptobyte_asn1.OCTET_STRING) {
+		return nil, errors.New("malformed primary key")
+	}
+	if !s.ReadASN1Bytes(&pk.Unique, cryptobyte_asn1.OCTET_STRING) {
+		return nil, errors.New("malformed unique key")
+	}
+
+	return pk, nil
+}
+
+func (k *protectedKeys) unlockKey(alg crypto.Hash) DiskUnlockKey {
+	if alg == crypto.Hash(nilHash) {
+		// This is to support the legacy TPM key data created
+		// via tpm2.NewKeyDataFromSealedKeyObjectFile.
+		return k.Unique
```

Comment:
this path is not exercised

Reply:
but if Version is 1 we don't arrive here,  as I said I'm confused by the role of KDFAlg vs Version here and the KeyParams, aynway if this is useful there should be test in keydata_test.go that shows how this is meant to happen

Reply:
This is exercised by the tests in the `tpm2` package - specifically, the ones in `tpm2/platform_legacy_test.go` that call `NewKeyDataFromSealedKeyObjectFile`. This creates an ephemeral v2 key data but with 0 for the `KDFAlg` parameter. This path exists just for that case.